chainer/training/extensions/parameter_statistics.py

```diff
@@ -0,0 +1,187 @@
+import numpy
+import six
+
+from chainer import cuda
+from chainer import reporter
+from chainer import training
+from chainer.training import extension
+
+
+def _iterable(x):
+    if isinstance(x, (list, tuple)):
+        return x
+    return x,
+
+
+def _prefix_dict_keys(prefix, x):
+    return {'{}/{}'.format(prefix, k): v for k, v in six.iteritems(x)}
+
+
+def _target_name(link, param_names, attr_names):
+    """Generate a dictionary key based on context."""
+    link_name = getattr(link, 'name', 'None')
+    param_name = '-'.join(_iterable(param_names))
+    attr_name = '-'.join(_iterable(attr_names))
+
+    return '{}/{}/{}'.format(link_name, param_name, attr_name)
+
+
+def _get_link_params(link, param_names, attr_names):
+    """Flatten link parameters into a single array and return a copy."""
+    params = []
+    for param in link.params():
+        if param.name in _iterable(param_names):
+            for attr_name in _iterable(attr_names):
+                params.append(getattr(param, attr_name).ravel())
+
+    return link.xp.concatenate(params) if params else link.xp.array([])
+
+
+def _statistics(x, functions):
+    """Compute statisticts for the given array.
+
+    Args:
+        x (array): Target array for which statistics are computed.
+        functions (iterable): Statistics to collect, mapping directly to NumPy
+            or CuPy functions.
+
+    Returns:
+        dict: Mappings from functions keys to statistic values.
+    """
+    stats = {}
+    for f in functions:
+        try:
+            # nan if x.size == 0 and f in ('mean', 'std')
+            stats[f] = getattr(x, f, lambda: float('NaN'))()
+        except ValueError:  # x.size == 0 and f in ('min, 'max')
+            stats[f] = float('NaN')
+
+    return stats
+
+
+def _percentiles(x, sigmas):
+    """Compute percentiles for the given array.
+
+    Args:
+        x (array): Target array for which percentiles are computed.
+        sigmas (iterable): Percentile sigma values.
+
+    Returns:
+        array: List of percentiles. The list has the same length as the given
+            ``sigma``.
+    """
+    def _percentiles_cpu(_x):
+        try:
+            return numpy.percentile(_x, sigmas).astype(_x.dtype)
+        except IndexError:  # _x.size == 0
+            return numpy.array((float('NaN'),) * len(sigmas))
+
+    if cuda.available and isinstance(x, cuda.ndarray):
+        # cuda.cupy.percentile() is not implemented
+        return cuda.to_gpu(_percentiles_cpu(cuda.to_cpu(x)))
+
+    return _percentiles_cpu(x)
+
+
+def _zeros(x):
+    """Count the number of zeros in the given array.
+
+    Args:
+        x (array): Target array for which sparsity is computed.
+
+    Returns:
+        int: Number of zeros.
+    """
+    if x.ndim == 0:
+        raise ValueError('Cannot count zeros for shape {}'.format(x.shape))
```

Comment:
Yes, that is probably better. Just noticed that links such as `BatchNormalization` contain parameters with `ndim == 0`. Fixed.